pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go

```diff
@@ -114,26 +119,36 @@ func (c *CSIDriverNodeServiceController) Name() string {
 func (c *CSIDriverNodeServiceController) sync(ctx context.Context, syncContext factory.SyncContext) error {
 	opSpec, opStatus, _, err := c.operatorClient.GetOperatorState()
 	if err != nil {
-		if apierrors.IsNotFound(err) {
-			return nil
-		}
 		return err
 	}
 
-	if opSpec.ManagementState != opv1.Managed {
+	action := management.GetSyncAction(opSpec.ManagementState)
+	klog.V(4).Infof("Syncing with managementState %q: %s", opSpec.ManagementState, action)
+	switch action {
+	case management.SyncActionManage:
+		return c.syncManaged(opSpec, opStatus, ctx, syncContext)
+	case management.SyncActionDelete:
+		return c.syncRemoved(opSpec, opStatus, ctx, syncContext)
+	case management.SyncActionIgnore:
 		return nil
+	default:
+		return fmt.Errorf("unrecognized managementState: %q: %s", opSpec.ManagementState, action)
 	}
+}
 
-	manifest := replacePlaceholders(c.manifest, opSpec)
-	required := resourceread.ReadDaemonSetV1OrDie(manifest)
-
-	for i := range c.optionalDaemonSetHooks {
-		err := c.optionalDaemonSetHooks[i](opSpec, required)
+func (c *CSIDriverNodeServiceController) syncManaged(opSpec *opv1.OperatorSpec, opStatus *opv1.OperatorStatus, ctx context.Context, syncContext factory.SyncContext) error {
+	if !management.IsOperatorNotRemovable() {
+		err := c.operatorClient.AddFinalizer(c.name)
```

Comment:
give me a comment here with an example finalizer please.

Reply:
Each library-go controller adds its own finalizer, and I used controller name for that. In EFS it looks like:

```
  finalizers:
  - CSIStaticResourceController
  - AWSEFSDriverNodeServiceController
```

Other removable CSI operators will have one more `*DriverControllerServiceController`. BTW, non-removabe operators / CSI drivers won't have any finalizers.